### PR TITLE
[css-anchor-position-1] Implement parsing `@position-try` rule

### PR DESCRIPTION
**Pull request by tuankiet65**
#### cc994137838b32d393c63fb549185375a96e2354
<pre>
[css-anchor-position-1] Implement parsing `@position-try` rule
<a href="https://rdar.apple.com/129789726">rdar://129789726</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=275455">https://bugs.webkit.org/show_bug.cgi?id=275455</a>

Reviewed by Antti Koivisto.

Implement parsing `@position-try` rule in CSS Anchor Positioning spec.
Note that the rules are only parsed and not yet collected for use in
`position-try-fallbacks`. This will be implemented in a future patch,
at least after parsing support for `position-try-fallbacks` is
completed.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-allowed-declarations-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-invalidation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-invalidation-shadow-dom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/at-position-try-parse-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-try-rule-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/SaferCPPExpectations/RefCntblBaseVirtualDtorExpectations:
  Added exception for CSSPositionTryRule, which inherits StyleRuleBase.
  StyleRuleBase doesn&apos;t define a virtual destructor, which trips up the
  checker. This is okay since StyleRuleBase defines its own delete operator
  that calls the correct destructor depends on the derived type. See
  <a href="https://github.com/WebKit/WebKit/pull/37285.">https://github.com/WebKit/WebKit/pull/37285.</a>
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
  Added exception for JSCSSPositionTryRule.cpp. Auto-generated file so can&apos;t
  be fixed except by fixing the code generator.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSRuleCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSPositionTryRule.cpp: Added.
(WebCore::StyleRulePositionTry::create):
(WebCore::StyleRulePositionTry::StyleRulePositionTry):
(WebCore::StyleRulePositionTry::protectedMutableProperties):
(WebCore::CSSPositionTryRule::create):
(WebCore::CSSPositionTryRule::CSSPositionTryRule):
(WebCore::CSSPositionTryRule::cssText const):
(WebCore::CSSPositionTryRule::reattach):
(WebCore::CSSPositionTryRule::name const):
(WebCore::CSSPositionTryRule::style):
* Source/WebCore/css/CSSPositionTryRule.h: Added.
(isType):
* Source/WebCore/css/CSSPositionTryRule.idl: Added.
* Source/WebCore/css/CSSProperty.cpp:
(WebCore::CSSProperty::isInsetProperty): Added some inset properties that we missed.
(WebCore::CSSProperty::isSizingProperty): Moved from AnchorPositionEvaluator.cpp
(WebCore::CSSProperty::isMarginProperty): Moved from AnchorPositionEvaluator.cpp
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRuleBase::visitDerived):
(WebCore::StyleRuleBase::createCSSOMWrapper const):
* Source/WebCore/css/StyleRule.h:
(WebCore::StyleRuleBase::isPositionTryRule const):
* Source/WebCore/css/StyleRuleType.h: Added new StyleRule type for @position-try.
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::traverseSubresources const):
(WebCore::StyleSheetContents::mayDependOnBaseURL const):
* Source/WebCore/css/parser/CSSAtRuleID.cpp:
(WebCore::cssAtRuleID):
* Source/WebCore/css/parser/CSSAtRuleID.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeAtRule): Added case to consume @position-try rule.
(WebCore::CSSParserImpl::consumePositionTryRule): Added code to parse @position-try.
(WebCore::ruleDoesNotAllowImportant):
(WebCore::CSSParserImpl::consumeDeclaration): Returned early if current style rule does not
allow !important. Rejected custom properties when parsing @position-try.
* Source/WebCore/css/parser/CSSParserImpl.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseValue): Added code to parse descriptors in @position-try rule.
(WebCore::propertyAllowedInPositionTryRule):
(WebCore::CSSPropertyParser::parsePositionTryDescriptor):
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/query/MediaQueryEvaluator.cpp:
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::AnchorPositionEvaluator::evaluateSize):
(WebCore::Style::isSizingProperty): Moved to CSSProperty.cpp
(WebCore::Style::isMarginProperty): Moved to CSSProperty.cpp
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::addChildRule):

Canonical link: <a href="https://commits.webkit.org/287528@main">https://commits.webkit.org/287528@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/1539c506b49394cd2e0d4cc93d114c5977b0f6ef

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79829 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58830 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33228 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84354 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30828 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81938 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67904 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7120 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62398 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20240 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82897 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52455 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72701 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42709 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49795 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26850 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29279 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27323 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85780 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7060 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4949 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70661 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7234 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68541 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69906 "70 api tests failed or timed out") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17448 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13908 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12823 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7019 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12592 "Found 3 new failures in JSWebCodecsEncodedVideoChunk.cpp, css/CSSPositionTryRule.h, JSCSSPositionTryRule.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6877 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10389 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8683 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/css/CSSPositionTryRule.cpp
Comment (nt1m) [inline Source/WebCore/css/CSSPositionTryRule.cpp]:
```suggestion
    builder.append("@position-try "_s, name(), " {"_s);
```

It's more efficient to do coalesce calls as much as possible.

File: Source/WebCore/css/CSSPositionTryRule.cpp
Comment (nt1m) [inline Source/WebCore/css/CSSPositionTryRule.cpp]:
```suggestion
        builder.append(' ', declarations, ' ');
```

File: Source/WebCore/css/CSSPositionTryRule.cpp
Comment (nt1m) [inline Source/WebCore/css/CSSPositionTryRule.cpp]:
```suggestion
        builder.append(' ');
```

File: Source/WebCore/css/CSSPositionTryRule.cpp
Comment (nt1m) [inline Source/WebCore/css/CSSPositionTryRule.cpp]:
```suggestion
    builder.append('}');
```

File: Source/WebCore/css/CSSPositionTryRule.h
Comment (nt1m) [inline Source/WebCore/css/CSSPositionTryRule.h]: nit: remove this.

File: Source/WebCore/css/CSSPositionTryRule.idl
Comment (nt1m) [inline Source/WebCore/css/CSSPositionTryRule.idl]: Can you add the bugzilla link before landing?

File: Source/WebCore/css/parser/CSSParserImpl.cpp
Comment (nt1m) [inline Source/WebCore/css/parser/CSSParserImpl.cpp]: ruleUsesAllowedRules ?

File: Source/WebCore/css/parser/CSSPropertyParser.cpp
Comment (nt1m) [inline Source/WebCore/css/parser/CSSPropertyParser.cpp]: Just use the bugzilla link for position-area?

File: Source/WebCore/style/RuleSetBuilder.cpp
Comment (nt1m) [inline Source/WebCore/style/RuleSetBuilder.cpp]: I would just remove this for now, and re-add it when you have an implementation of the evaluation ready.

File: Source/WebCore/css/parser/CSSParserImpl.h
Comment (nt1m) [inline Source/WebCore/css/parser/CSSParserImpl.h]: CounterStyle/ViewTransition/PositionTry should actually just use `NoRules` since you can't nest further rules inside them. It is not the case however for the other ones.

File: Source/WebCore/css/CSSPositionTryRule.cpp
Comment (tuankiet65) [reply Source/WebCore/css/CSSPositionTryRule.cpp]: I didn't know append can take an arbitrary number of arguments!

File: Source/WebCore/css/CSSPositionTryRule.h
Comment (tuankiet65) [reply Source/WebCore/css/CSSPositionTryRule.h]: Yeah I had StyleRuleViewTransition there as a template to follow but forgot to remove it.

File: Source/WebCore/css/parser/CSSParserImpl.h
Comment (tuankiet65) [reply Source/WebCore/css/parser/CSSParserImpl.h]: I decided to do some cleanup: https://github.com/WebKit/WebKit/pull/37163

File: Source/WebCore/css/parser/CSSParserImpl.cpp
Comment (anttijk) [reply Source/WebCore/css/parser/CSSParserImpl.cpp]: A general principle is to name booleans starting with a yes/no verb, like `isUnrestrictedRule`.

A switch might be better here.

File: Source/WebCore/css/parser/CSSParserImpl.cpp
Comment (anttijk) [inline Source/WebCore/css/parser/CSSParserImpl.cpp]: Something used for interacting with Inspector though details are hazy to me. (please don't land this FIXME)

**Review by anttijk — APPROVED** [review]

File: Source/WebCore/css/CSSPositionTryRule.h
Comment (nt1m) [inline Source/WebCore/css/CSSPositionTryRule.h]: You can probably drop this FIXME?

File: Source/WebCore/css/parser/CSSParserImpl.cpp
Comment (nt1m) [reply Source/WebCore/css/parser/CSSParserImpl.cpp]: Unrestricted is probably not the right verb fwiw, `allowedRules` are a set of restrictions too. Maybe just drop the variable and use a switch as Antti says.

File: Source/WebCore/css/parser/CSSParserImpl.cpp
Comment (nt1m) [inline Source/WebCore/css/parser/CSSParserImpl.cpp]:
```suggestion
    // Prelude should only be a <dashed-ident>.
```

File: Source/WebCore/css/parser/CSSParserImpl.cpp
Comment (nt1m) [inline Source/WebCore/css/parser/CSSParserImpl.cpp]: You can't use !important in position-try from what I can understand, so probably makes sense to me to add it in there.

File: Source/WebCore/inspector/InspectorStyleSheet.cpp
Comment (nt1m) [inline Source/WebCore/inspector/InspectorStyleSheet.cpp]: Can you check if this works and remove the FIXME as appropriate? If it doesn't then please create a bug for it. If you don't know, I'd rather not include it.

File: Source/WebCore/style/RuleSetBuilder.cpp
Comment (nt1m) [reply Source/WebCore/style/RuleSetBuilder.cpp]: I'd remove this for now.

File: Source/WebCore/css/parser/CSSParserImpl.cpp
Comment (tuankiet65) [reply Source/WebCore/css/parser/CSSParserImpl.cpp]: 👍

**Comment by tuankiet65** [comment]
Currently blocked-ish by https://github.com/WebKit/WebKit/pull/37285 - mac-safer-cpp pointed out `StyleRuleBase` doesn't have a virtual destructor.

File: Source/WebCore/css/StyleRule.h
Comment (mdubet) [inline Source/WebCore/css/StyleRule.h]: We should keep `StyleRuleBase` non virtual, and implement the trait system for it (cf the functions at the end of StyleRule.h) 

```
SPECIALIZE_TYPE_TRAITS_BEGIN(WebCore::StyleRulePositionTry)
    static bool isType(const WebCore::StyleRuleBase& rule) { return rule.isPositionTry(); }
SPECIALIZE_TYPE_TRAITS_END()
```

Comment (tuankiet65) [reply Source/WebCore/css/StyleRule.h]: Right, this is before I know `StyleRuleBase` is intentionally designed to be non-virtual

File: Source/WebCore/css/parser/CSSParserImpl.cpp
Comment (tuankiet65) [reply Source/WebCore/css/parser/CSSParserImpl.cpp]: We dropped some `AllowRules` enums so the `if` condition is shorter now, hence I'll just keep the `if`. But yeah, I might make it into a switch if I have time.

File: Source/WebCore/css/StyleRule.h
Comment (tuankiet65) [reply Source/WebCore/css/StyleRule.h]: I changed it back to the original.

File: Source/WebCore/css/parser/CSSParserImpl.cpp
Comment (tuankiet65) [reply Source/WebCore/css/parser/CSSParserImpl.cpp]: I added it there.

File: Source/WebCore/inspector/InspectorStyleSheet.cpp
Comment (tuankiet65) [reply Source/WebCore/inspector/InspectorStyleSheet.cpp]: https://bugs.webkit.org/show_bug.cgi?id=284176

File: Source/WebCore/style/RuleSetBuilder.cpp
Comment (tuankiet65) [reply Source/WebCore/style/RuleSetBuilder.cpp]: I removed it.

**Pull request merged**
